lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -308,48 +339,36 @@ impl ManagerInner {
-        };
-
-        // The overlay contents may be stored in the table itself (for pending
-        // overlays) or in guest memory (for active overlays). Before bothering
-        // to obtain them, create a mapping for the target page to verify that
-        // the target PFN is valid. This mapping will be used later if this is
-        // the first overlay for the target PFN.
+        // Moving an overlay consists of applying it to its new location and
+        // then removing it from its old one. This is only legal if the target
+        // PFN is valid and there is no other overlay of this kind associated
+        // with that PFN.
+        //
+        // Checking these conditions up front allows this function to attempt to
+        // remove the existing overlay (checking for errors) and then assert
+        // that the overlay can be applied in its new position. This is simpler
+        // than removing the old overlay, failing to add the new one, and then
+        // having to remember how to add the old overlay back in its previous
+        // position.
         let mut to_mapping = to_pfn.map(&memctx)?;
+        if let Some(to_set) = self.overlays.get(&to_pfn) {
+            if to_set.contains_kind(kind) {
+                return Err(OverlayError::KindInUse(to_pfn.into(), kind));
+            }
+        }
 
-        // Get the page contents to transfer, but don't actually change the
-        // tables yet, since applying the new overlay might still fail if there
-        // is an existing overlay of this kind at the target PFN.
-        //
-        // It's possible to do this check now, modify the tables in-place, and
-        // then assert that adding the new overlay succeeded, but this sequence
-        // allows this function to reuse the existing overlay removal code. This
-        // is unlikely to be a bottleneck in practice because guests change
-        // their overlay locations so infrequently (and any guest that spends
-        // all its time moving overlays around has, in some ways, asked for
-        // it...).
-        let contents = if from_set.active == kind {
-            let mut from_mapping = from_pfn
-                .map(&memctx)
-                .expect("active overlay PFNs can be mapped");
-            let mut contents = OverlayContents::default();
-            from_mapping.read_page(&mut contents);
-            contents
-        } else if let Some(pending_contents) = from_set.pending.get(&kind) {
-            pending_contents.clone()
-        } else {
-            return Err(OverlayError::NoOverlayOfKind(from_pfn.into(), kind));
-        };
+        // Removing the old overlay might fail if it doesn't exist in
+        // `from_pfn`, but if it succeeds it will always produce the previous
+        // page contents.
+        let contents = self
+            .remove_overlay(from_pfn, kind, acc_mem, RemoveReturnContents::Yes)?
+            .expect("asked for old contents");
```

Comment:
having to `expect` this is a bit of a bummer, but overall, it still seems less grody than the previous thing with `MaybeUninit`...

Reply:
Yeah, I don't think there's really a better way out of this one, but at least there's no `unsafe`.